cdap-docs/tools/doc-cdap-default.py

```diff
@@ -0,0 +1,434 @@
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+# Builds an rst table to document the cdap-default.xml file
+# Any property names listed in cdap-default-exclusions.txt are excluded.
+# Any properties marked "final" are placed in second table after the first.
+#
+# python doc-cdap-default.py <source-filepath> <output-filepath>
+# python doc-cdap-default.py <output-filepath>
+#
+# If <source-filepath> not provided, uses CDAP_DEFAULT_XML
+# if <output-filepath> not provided, outputs to standard out
+
+
+# Steps:
+#
+# Load source code cdap-default.xml
+# keys: name of property
+# values: Property class object (value and description)
+# Defaults in source code are in no particular order.
+# 
+# Sort by key
+# Output rst table to standard out
+# If filepath provided, writes to file instead
+
+import os
+import sys
+import textwrap
+import xml.etree.ElementTree as ET
+
+from datetime import datetime
+
+RELATIVE_PATH = '../..'
+CDAP_DEFAULT_XML = 'cdap-common/src/main/resources/cdap-default.xml'
+CDAP_DEFAULT_EXCLUSIONS = 'cdap-default-exclusions.txt'
+
+FIRST_TWO_SECTIONS = ['General Configuration', 'Global Configuration']
+
+
+RST_TABLE_HEADER = """
+.. list-table::
+   :widths: 30 35 35
+   :header-rows: 1
+
+   * - Parameter Name
+     - Default Value
+     - Description"""
+
+RST_TABLE_HEADER_FINAL = """
+
+Final Configurations
+--------------------
+These properties are marked as ``final``, meaning that their value cannot be changed, even
+with a setting in the ``cdap-site.xml``:
+""" + RST_TABLE_HEADER + '\n'
+
+NAME_START  = '   * - '
+VALUE_START = '     - '
+BLOCK_START = '| ``'
+BLOCK_JOIN  = '``\n       | ``'
+DESC_START  = '     - '
+
+SECTION_START = NAME_START + '|\n       | '
+
+# The XML_HEADER is not ascii text due to the copyright symbol in the fourth line
+
+XML_HEADER = u"""<?xml version="1.0"?>
+<?xml-stylesheet type="text/xsl" href="configuration.xsl"?>
+<!--
+  Copyright © 2014-%d Cask Data, Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License"); you may not
+  use this file except in compliance with the License. You may obtain a copy of
+  the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+  License for the specific language governing permissions and limitations under
+  the License.
+  -->
+""" % datetime.now().year
+
+
+class Item:
```

Comment:
In this file, should we make the tab indent to be 2 spaces instead of 4 to be consistent with all our code?